### ISSUE TITLE
Aggregate result of Hover participants

### ISSUE THREAD

**Issue by Seiphon** [angelozerr/lsp4xml#577]
Hi,

I encountered a problem when I try to implement hover feature on a extension plugin. Take WebPlugin for an example. I create a WebHoverParticipant class which extends HoverParticipantAdapter, then implement my hover feature by it. But it never works.

After I debugging the code, I found the the reason which cause that issue.
org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLHover.java
getTagHover, getAttrNameHover and getAttrValueHover methods has code like follows:

`

for (IHoverParticipant participant : extensionsRegistry.getHoverParticipants()) {

    try {
        Hover hover = participant.onTag(hoverRequest);
            if (hover != null) {
                return hover;
            }
    } catch (Exception e) {
    }
}
`

After get all participants, ContentModelHoverParticipant has a higher priority. when this participant get a hover which is not null, this method will return. So the onTag method of rest of participants will not be called any more.

I am wondering is that a bug? if it's not a bug, and the code was designed like this on purpose by some reason. can you explain that reason?

Thank you guys, hope get your reply.

Seiphon

**Comment by fbricon** [comment]
Looks like an oversight, we might have similar problems with other extension participants.

**Comment by angelozerr** [comment]
> I am wondering is that a bug? if it's not a bug, and the code was designed like this on purpose by some reason. can you explain that reason?

To be honest with you, we try to give the capability to extend LSP4XML with SPI and we use this feature with our internal support (XSD, DTD, etc).

Your use-case shows that there is a limitation with hover design and I'm totally aware to change that. I think we should refactor hover participant to give the capability to aggregate hover result and not return the first hover which is not null.

Today  `IHoverParticipant ` looks like this:

```java
public interface IHoverParticipant {

	Hover onTag(IHoverRequest request) throws Exception;

	Hover onAttributeName(IHoverRequest request) throws Exception;

	Hover onAttributeValue(IHoverRequest request) throws Exception;

}
```

Today we use `MarkupContent` in the Hover result, but if we can use  `MarkedString` (I don't understand very well the difference for the moment and I don't know if there are a capability to use `MarkedString`) we could replace `Hover` with  `MarkedString` 


Once all participant return several  `MarkedString` , we could create an Hover instance with this constructor:

```java
public Hover(@NonNull final List<Either<String, MarkedString>> contents, final Range range) 
```

@fbricon what do you think about that?

@Seiphon which LSP client are you intention to use it? vscode-xml, LSP4E with Eclipse with Wild Deb Developer, other? I ask you that because today It seems none LSP client could support the LSP4XML extension.

For vscode, I plan to give this capability, see https://github.com/redhat-developer/vscode-xml/issues/190

**Comment by angelozerr** [comment]
@Seiphon if you can and want to create a PR by following the idea of `MarkedString`, please do it! Thanks!

**Comment by angelozerr** [comment]
@Seiphon  please wait, it seems MarkedString is deprecated and MarkupContent should be used. Please wait that I see what we must to do with @fbricon

**Comment by Seiphon** [comment]
> @Seiphon which LSP client are you intention to use it? vscode-xml, LSP4E with Eclipse with Wild Deb Developer, other? I ask you that because today It seems none LSP client could support the LSP4XML extension.

Hi @angelozerr , I work for Liferay Inc. We are trying to develop a LSP to support some liferay feature staff, and we did some research on that recently. I found LSP4XML is supper cool to support LSP features for xml files. So I try to use LSP4XML extension to implement hover, code proposal, validation features for some liferay-style xml files. I used LSP4E with Eclipse with my own LSP client for test for now. If one day wild web developer can support the LSP4XML extensioin, it would be great!

If there is something I can do, just comment me, I will try my best to help.

Thank you so much

**Comment by angelozerr** [comment]
> I work for Liferay Inc. We are trying to develop a LSP to support some liferay feature staff, and we did some research on that recently. I found LSP4XML is supper cool to support LSP features for xml files. 

Thanks!

> So I try to use LSP4XML extension to implement hover, code proposal, validation features for some liferay-style xml files. I used LSP4E with Eclipse with my own LSP client for test for now. 

Ok I understand and I think you work with @jtydhr88 . We are aware of this (client) extension problem:

 * for extending LSP4XML in vscode, see https://github.com/redhat-developer/vscode-xml/issues/190
 * for extending LSP4XML in  wildwebdeveloper, see https://github.com/eclipse/wildwebdeveloper/issues/266

> If one day wild web developer can support the LSP4XML extensioin, it would be great!

I would like too! See https://github.com/eclipse/wildwebdeveloper/issues/266

> If there is something I can do, just comment me, I will try my best to help.

After discussion with @fbricon I think participant should return just a String formatted with proper format (plaintext, markdown according the client support).

In other words, we would like to provide this API:

```
public interface IHoverParticipant {

	String onTag(IHoverRequest request) throws Exception;

	String onAttributeName(IHoverRequest request) throws Exception;

	String onAttributeValue(IHoverRequest request) throws Exception;

}
```

The idea is to loop for each String particpant and check if it's not null. In this case, aggregration will be done and MarkupContent will be returned with the aggregate String participant.

I don't know if it's a simple task. We have a lot of tests with hover so I'm confident to break the API and be sure that we will not break something.

It's not in our priority now, so if you wish to do something, please create a PR otherwise please wait for we implement this feature.

> Thank you so much

**Comment by Seiphon** [comment]
Hi @angelozerr  I have sent a PR for this issue according to my understanding. If something is incorrect here, please let me know and I will resend the PR.
https://github.com/angelozerr/lsp4xml/pull/582

**Comment by angelozerr** [comment]
Thanks @Seiphon !

> If something is incorrect here, please let me know and I will resend the PR.

I see quickly, and it follows the idea that we have in mind.  Let continue our discussion with comments in your PR.

**Comment by angelozerr** [comment]
@Seiphon I review your PR which is very good. Please see my comments. We need tests too with aggregate support.

**Comment by angelozerr** [comment]
Fixed with https://github.com/angelozerr/lsp4xml/pull/582
